shesmu-server-ui/src/simulation.ts

```diff
@@ -0,0 +1,781 @@
+                      string,
+                      (input: any[]) => any
+                    ]
+                )
+              )
+            : "No records provided to dumper."
+        );
+        tabList.push({
+          name: "Dumpers",
+          contents: group(
+            dropdown(
+              (input) => (input ? input[0] : blank()),
+              Object.entries(response.dumpers)[0],
+              dumpState.model,
+              null,
+              ...Object.entries(response.dumpers)
+            ),
+            br(),
+            dumpState.ui
+          ),
+        });
+      }
+      if (response.overloadedInputs?.length) {
+        tabList.push({
+          name: "Overloaded Inputs",
+          contents: [
+            text(
+              "The following input formats are unavailable and prevented the simulation from running:"
+            ),
+            br(),
+            table(response.overloadedInputs, ["Input Format", (x) => x]),
+          ],
+        });
+      }
+
+      if (response.overloadedServices?.length) {
+        tabList.push({
+          name: "Overloaded Services",
+          contents: [
+            text(
+              "The following services are unavailable and prevented the simulation from running:"
+            ),
+            br(),
+            table(response.overloadedServices, ["Service", (x) => x]),
+          ],
+        });
+      }
+      if (response.metrics) {
+        tabList.push({
+          name: "Prometheus Metrics",
+          contents: preformatted(response.metrics),
+        });
+      }
+      if (response.bytecode) {
+        tabList.push({
+          name: "Bytecode",
+          contents: preformatted(response.bytecode),
+        });
+      }
+    }
+    const dashboardTabs = tabs(...tabList);
+    setFindHandler(dashboardTabs.find);
+    return dashboardTabs.ui;
+  });
+  const main = refreshable(
+    "/simulate",
+    (request) => {
+      editor.getSession().clearAnnotations();
+      return {
+        body: JSON.stringify({
+          fakeActions: Object.fromEntries(fakeActionDefinitions),
+          dryRun: false,
+          readStale: !waitForData.getter(),
+          script: request,
+        }),
+        method: "POST",
+      };
+    },
+    combineModels(dashboardState.model, errorTable.model)
+  );
+
+  const { last: lastTheme, model: savedTheme } = locallyStoredString(
+    "shesmu_theme",
+    "ace/theme/chrome"
+  );
+  addElements(
+    container,
+    group(
+      button("🤖 Simulate", "Run olive simulation and fetch results", () =>
+        main.statusChanged(editor.getValue())
+      ),
+      waitForData.ui,
+      buttonAccessory(
+        "🠝 Upload File",
+        "Upload a file from your computer to simulate",
+        () =>
+          loadFile((name, data) => {
+            fileName = name;
+            editor.setValue(data, 0);
+          })
+      ),
+      buttonAccessory(
+        "🡇 Download File",
+        "Save this editor to your computer",
```

Comment:
Saving 'this editor' sounds a little like it'll download a whole executable or something, which I take it it's not doing?

Reply:
No, but that would be cool. ;-)